### PR TITLE
feat: add options option to structures

### PR DESCRIPTION
**Pull request by SebestikCZ**
**Please describe the changes this PR makes and why it should be merged:**

**Status and versioning classification:**
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

**Comment by S222em** [comment]
I am not sure if this should be added, as users can just extend the Command class for this feature, and with the new slash commands I don't see it getting used anyways. I'll let @xHyroM decide on this

**Comment by xhyrom** [comment]
Option 1

I agree with @S222em that people can extend the command like I have in my bot for example:
```ts
import { Command, CommandOptions } from 'gcommands';

export type CommandCategory = 'management' | 'moderation' | 'utility' | 'developer';
export type ExtendedCommandOptions = CommandOptions & {
    syntax?: string;
    category?: CommandCategory;
}

export class ExtendedCommand extends Command {
	syntax?: string;
	category?: CommandCategory;
	constructor(options: ExtendedCommandOptions) {
		super(options);

		this.syntax = options.syntax;
		this.category = options.category;
	}
}
```

We could add this example to the guide. 

Option 2
Maybe we could still make some raw proper ones `options` or `data` where all the raw options you put in the command would be.

**Comment by S222em** [comment]
> Option 1
> 
> 
> 
> I agree with @S222em that people can extend the command like I have in my bot for example:
> 
> ```ts
> 
> import { Command, CommandOptions } from 'gcommands';
> 
> 
> 
> export type CommandCategory = 'management' | 'moderation' | 'utility' | 'developer';
> 
> export type ExtendedCommandOptions = CommandOptions & {
> 
>     syntax?: string;
> 
>     category?: CommandCategory;
> 
> }
> 
> 
> 
> export class ExtendedCommand extends Command {
> 
> 	syntax?: string;
> 
> 	category?: CommandCategory;
> 
> 	constructor(options: ExtendedCommandOptions) {
> 
> 		super(options);
> 
> 
> 
> 		this.syntax = options.syntax;
> 
> 		this.category = options.category;
> 
> 	}
> 
> }
> 
> ```
> 
> 
> 
> We could add this example to the guide. 
> 
> 
> 
> Option 2
> 
> Maybe we could still make some raw proper ones `options` or `data` where all the raw options you put in the command would be.

Agreed. @SebestikCZ can you do what hyro suggested in option 2?

**Comment by SebestikCZ** [comment]
> > Option 1
> > 
> > 
> > 
> > I agree with @S222em that people can extend the command like I have in my bot for example:
> > 
> > ```ts
> > 
> > import { Command, CommandOptions } from 'gcommands';
> > 
> > 
> > 
> > export type CommandCategory = 'management' | 'moderation' | 'utility' | 'developer';
> > 
> > export type ExtendedCommandOptions = CommandOptions & {
> > 
> >     syntax?: string;
> > 
> >     category?: CommandCategory;
> > 
> > }
> > 
> > 
> > 
> > export class ExtendedCommand extends Command {
> > 
> > 	syntax?: string;
> > 
> > 	category?: CommandCategory;
> > 
> > 	constructor(options: ExtendedCommandOptions) {
> > 
> > 		super(options);
> > 
> > 
> > 
> > 		this.syntax = options.syntax;
> > 
> > 		this.category = options.category;
> > 
> > 	}
> > 
> > }
> > 
> > ```
> > 
> > 
> > 
> > We could add this example to the guide. 
> > 
> > 
> > 
> > Option 2
> > 
> > Maybe we could still make some raw proper ones `options` or `data` where all the raw options you put in the command would be.
> 
> Agreed. @SebestikCZ can you do what hyro suggested in option 2?

I could do that, but I don't know, what do you want from me _almost where..._

**Comment by S222em** [comment]
@SebestikCZ as hyro suggested in option 2, add a Command#options field (for Components, Plugins and Listeners as well please) where the options for the command are stored, so instead of command.name, it would become command.options.name

Discord.js also uses this in their Client class for example

**Comment by SebestikCZ** [comment]
okay, seems legit to make

**Comment by xhyrom** [comment]
> @SebestikCZ as hyro suggested in option 2, add a Command#options field (for Components, Plugins and Listeners as well please) where the options for the command are stored, so instead of command.name, it would become command.options.name
> 
> Discord.js also uses this in their Client class for example

But we need keep this.name (example) for minor change

Just set this.options after constructor and then valid it with zod and set this.name for example

**Comment by SebestikCZ** [comment]
I comitted it to another branch reeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee
_once again_

**Comment by SebestikCZ** [comment]
hm, a little change will be needed

File: src/lib/structures/Command.ts [hunk 1]
```diff
@@ -160,6 +161,7 @@ export class Command {
 				this.run = options.run || Command.defaults?.run;
 				this.onError = options.onError || Command.defaults?.onError;
 				this.autoDefer = options.autoDefer || Command.defaults?.autoDefer;
+				this.options = options;
```
Comment (xhyrom) [inline src/lib/structures/Command.ts]: Put this before zod validation (same for listener)

https://github.com/SebestikCZ/GCommands/blob/c890545d3ce22de37df900aaa2f86db730e89455/src/lib/structures/Command.ts#L135-L136

**Review by xhyrom — CHANGES_REQUESTED** [review]

**Comment by S222em** [comment]
@xHyroM what about removing all the options on the command level? e.g. remove `command.name` etc, so options will only reside in the `command.options` and the run method can be replaced with one that catches any errors itself, instead of doing that in the handler.

**Comment by xhyrom** [comment]
> @xHyroM what about removing all the options on the command level? e.g. remove `command.name` etc, so options will only reside in the `command.options` and the run method can be replaced with one that catches any errors itself, instead of doing that in the handler.

hm, idk we want minor change now

**Comment by S222em** [comment]
> > @SebestikCZ as hyro suggested in option 2, add a Command#options field (for Components, Plugins and Listeners as well please) where the options for the command are stored, so instead of command.name, it would become command.options.name
> > Discord.js also uses this in their Client class for example
> 
> But we need keep this.name (example) for minor change
> 
> Just set this.options after constructor and then valid it with zod and set this.name for example

@xHyroM shouldn't `this.options` be set after zod validated it? Zod actually transforms the value on some cases as well, and that wont be reflected into `this.options` then. Zod is also set to let any additional items trough so there should be no issue.

**Comment by xhyrom** [comment]
> > > @SebestikCZ as hyro suggested in option 2, add a Command#options field (for Components, Plugins and Listeners as well please) where the options for the command are stored, so instead of command.name, it would become command.options.name
> > > Discord.js also uses this in their Client class for example
> > 
> > 
> > But we need keep this.name (example) for minor change
> > Just set this.options after constructor and then valid it with zod and set this.name for example
> 
> @xHyroM shouldn't `this.options` be set after zod validated it? Zod actually transforms the value on some cases as well, and that wont be reflected into `this.options` then. Zod is also set to let any additional items trough so there should be no issue.

ah, okay then @SebestikCZ time to revert it xd

File: src/lib/structures/Command.ts [hunk 2]
```diff
@@ -160,6 +160,7 @@ export class Command {
 				this.run = options.run || Command.defaults?.run;
 				this.onError = options.onError || Command.defaults?.onError;
 				this.autoDefer = options.autoDefer || Command.defaults?.autoDefer;
+				this.options = options || Command.defaults?;
```
Comment (S222em) [inline src/lib/structures/Command.ts]: You should merge the defaults with the options (options overwrites defaults). Be careful to not overwrite the defaults with the merged one.

Something like this should work:
```ts
this.options = { ...Command.defaults, ...options }
```

File: src/lib/structures/Component.ts [hunk 3]
```diff
@@ -90,6 +91,7 @@ export class Component {
 				this.run = options.run || Component.defaults?.run;
 				this.onError = options.onError || Component.defaults?.onError;
 				this.autoDefer = options.autoDefer || Component.defaults?.autoDefer;
+				this.options = options || Component.defaults?;
```
Comment (S222em) [inline src/lib/structures/Component.ts]: Same here

File: src/lib/structures/Plugin.ts [hunk 4]
```diff
@@ -13,13 +13,15 @@ const validationSchema = z
 export class Plugin {
 	public name: string;
 	public run: (client: GClient) => any;
+	public options: Object;
```
Comment (S222em) [inline src/lib/structures/Plugin.ts]: This should be typed beter

**Review by S222em — CHANGES_REQUESTED** [review]

**Comment by xhyrom** [comment]
@SebestikCZ

**Comment by SebestikCZ** [comment]
If you have ideas, how to make it right, set up commits and I will add it to this pull

**Comment by xhyrom** [comment]
> If you have ideas, how to make it right, set up commits and I will add it to this pull

Just type `options` in Plugin.ts

```ts
export interface PluginOptions {
    name: string;
    run: (client: GClient) => any;
}
```

**Review by S222em — APPROVED** [review]

**Review by xhyrom — APPROVED** [review]

**Pull request merged**